### PR TITLE
Union types

### PR DESCRIPTION
**Pull request by gwprice115**

### CODE REVIEW

File: templates/base/client/.rtag/types.ts.hbs [hunk 1]
```diff
@@ -5,6 +5,22 @@ export enum {{@key}} {
   {{label}},
   {{/each}}
 }
+{{else if (eq type "union")}}
+{{#each options}}
+interface {{@../key}}_{{> renderArg}} {
+  type: "{{> renderArg}}"
+  value: {{> renderArg}}
+}
+{{/each}}
```
Comment (hpx7) [inline templates/base/client/.rtag/types.ts.hbs]: why is this necessary?

File: templates/base/client/.rtag/types.ts.hbs [hunk 2]
```diff
@@ -5,6 +5,22 @@ export enum {{@key}} {
...
+export type {{@key}} =
+{{#each options}}
+ | {{@../key}}_{{> renderArg}}
```
Comment (hpx7) [inline templates/base/client/.rtag/types.ts.hbs]: can't we just do `| {{> renderArg}}` directly?

File: generate.ts [hunk 3]
```diff
@@ -119,6 +125,15 @@ function generate() {
 
   function getArgsInfo(args: z.infer<typeof TypeArgs>, required: boolean, alias: boolean, typeString?: string): Arg {
     if (Array.isArray(args)) {
+      const isUnion = args.every(arg => arg in doc.types);
+      if(isUnion) {
```
Comment (hpx7) [inline generate.ts]: nit: i think the above check can be inlined here

File: templates/base/client/.rtag/types.ts.hbs (hunk 1 above)
Comment (gwprice115) [reply templates/base/client/.rtag/types.ts.hbs]: we need the "type" string literal in order to write the type guards. another option that just occurred to me is we could write these as

`type Fruit = (Apple & {type: "Apple"}) | (Orange & {type: "Orange"}` which might be slightly friendlier. what do you think?

File: templates/base/client/.rtag/types.ts.hbs (hunk 2 above)
Comment (gwprice115) [reply templates/base/client/.rtag/types.ts.hbs]: no because we need to put the type string literal on there

File: generate.ts (hunk 3 above)
Comment (gwprice115) [reply generate.ts]: you mean not store the `isUnion` const and put the condition right in the `if`?

File: templates/base/client/.rtag/types.ts.hbs (hunk 2 above)
Comment (gwprice115) [reply templates/base/client/.rtag/types.ts.hbs]: per prior comment maybe there's a better way

File: generate.ts (hunk 3 above)
Comment (hpx7) [reply generate.ts]: yes

File: generate.ts [hunk 4]
```diff
@@ -24,13 +24,28 @@ const RtagConfig = z
...
   type: "object";
   alias: boolean;
   typeString?: string;
   properties: Record<string, Arg>;
 }
+interface UnionArg {
+  type: "union";
+  alias: boolean;
+  typeString?: string;
+  options: Record<string, Arg>;
```
Comment (gwprice115) [inline generate.ts]: why not object?

File: templates/base/client/.rtag/app.ts.hbs [hunk 5]
```diff
@@ -164,6 +164,38 @@ Vue.component("array-input", {
...
+      this.$emit("update", this.arg, value, valid);
+    },
+    updateType(value: string | null) {
+      const val = value === null ? undefined : { type: value };
+      this.$emit("update", this.arg, val, !this.required || val !== undefined);
+    },
+  },
+  computed: {
+    options() {
+      return Object.keys(Types[this.union as keyof typeof Types]).filter((x) => !x.startsWith("is"));
```
Comment (gwprice115) [inline templates/base/client/.rtag/app.ts.hbs]: lel

File: templates/base/client/.rtag/index.html.hbs [hunk 6]
```diff
@@ -66,6 +66,14 @@
 {{#with item}}
 {{> renderForm arg=../arg required=false}}
 {{/with}}
+{{else if (eq type "union")}}
+<union-input {{#if arg}}arg="{{arg}}" :value="value.{{arg}}"{{else}}:arg="arg" :value="value"{{/if}} :required="{{required}}" @update="update" union="{{typeString}}">
+  {{#each options}}
+  <template #{{@key}}="{ arg, value, update }">
+    {{> renderForm required=true}}
```
Comment (gwprice115) [inline templates/base/client/.rtag/index.html.hbs]: is this the recursion?

Comment (hpx7) [reply templates/base/client/.rtag/index.html.hbs]: yup

File: generate.ts (hunk 4 above)
Comment (hpx7) [reply generate.ts]: not sure what you're asking
